File: LeetCode/00-legacy-warmups/convert-sorted-array-to-binary-search-tree/solutions/main.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

var nums []int = []int{-10, -3, 0, 5, 9}

func main() {
	fmt.Println(sortedArrayToBST(nums))
}

func sortedArrayToBST(nums []int) *TreeNode {
	var treeResult TreeNode

	// 1. find root
	var rootIndex int = len(nums) / 2

	treeResult.Val = nums[rootIndex]

	if rootIndex+1 < len(nums) {
		treeResult.Right = sortedArrayToBST(nums[rootIndex+1:])

	}

	if rootIndex-1 > 0 {

		treeResult.Left = sortedArrayToBST(nums[0 : rootIndex])
	}

	return &treeResult
}

File: LeetCode/00-legacy-warmups/convert-sorted-array-to-binary-search-tree/solutions/main.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

const sortedArrayToBST = function (nums) {
  const rootIndex = Math.floor(nums.length / 2);
  const root = new TreeNode(nums[rootIndex]);

  if (rootIndex - 1 >= 0)
    root.left = sortedArrayToBST(nums.slice(0, rootIndex));
  if (rootIndex + 1 <= nums.length)
    root.right = sortedArrayToBST(nums.slice(rootIndex + 1));

  return root;
};